plugin/route53/route53.go

```diff
@@ -15,98 +20,180 @@ import (
 	"github.com/miekg/dns"
 )
 
-// Route53 is a plugin that returns RR from AWS route53
+// Route53 is a plugin that returns RR from AWS route53.
 type Route53 struct {
 	Next plugin.Handler
 
-	zones  []string
-	keys   map[string]string
-	client route53iface.Route53API
+	zoneNames []string
+	client    route53iface.Route53API
+	upstream  *upstream.Upstream
+
+	zMu   sync.RWMutex
+	zones map[string]*zone
 }
 
-// ServeDNS implements the plugin.Handler interface.
-func (rr Route53) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dns.Msg) (int, error) {
-	state := request.Request{W: w, Req: r}
-	qname := state.Name()
+type zone struct {
+	id string
+	z  *file.Zone
+}
 
-	zone := plugin.Zones(rr.zones).Matches(qname)
-	if zone == "" {
-		return plugin.NextOrFailure(rr.Name(), rr.Next, ctx, w, r)
+// New returns new *Route53.
+func New(
+	ctx context.Context,
+	c route53iface.Route53API,
+	keys map[string]string,
+	up *upstream.Upstream) (*Route53, error) {
```

Comment:
please put these on one line (there is no line length limit in go)

Reply:
Done.